[1단계 - 로또(자동)]수달(허수진) 미션 제출합니다.

her0807 opened the pull request:
## 인사
안녕하세요 핀 ,  수달입니다. 😵‍💫 
이번 미션동안 핀에게 많이 배우고 싶습니다. 
곧 3월 이지만 날씨는 겨울이네요. 감기 조심하세요. ㅎㅎ

미션을 진행하며 궁금했던 점들을 코멘트로 적어두었습니다.
감사합니다. 

- [View 에서 객체를 반환하는 것에 대하여](https://github.com/woowacourse/java-lotto/pull/381#discussion_r813643385)
- [싱글톤과 유틸 클래스의 차이](https://github.com/woowacourse/java-lotto/pull/381#discussion_r813649628)
- [수익률 계산이 구입금액의 역할인가?](https://github.com/woowacourse/java-lotto/pull/381#discussion_r813652444)

her0807 committed:
docs: 요구사항 도출

her0807 committed:
feat : Payment 검증 로직 작성

her0807 committed:
feat : LottoNumber 검증 로직 작성

her0807 committed:
feat : Lotto 검증 로직 작성

her0807 committed:
feat : WinningLotto  검증 로직 작성

her0807 committed:
docs: lottoNumber 요구사항 변경

her0807 committed:
feat : Input 로직 작성

her0807 committed:
docs : 출력에 관한 요구사항 추가

her0807 committed:
feat : 로또 오름차순으로 정렬

her0807 committed:
feat : 로또 생성 기능 추가

her0807 committed:
feat : 구입 갯수만큼 로또 자동 생성 기능 추가

kimchan123 committed:
feat : 당첨 등수 계산 로직 작성

kimchan123 committed:
feat : 당첨 통계 산출 로직 작성

her0807 committed:
feat : 출력 로직 작성

her0807 committed:
feat : 로또 전반적인 진행을 위한 controller 작성

her0807 committed:
fix : 로또 생성 시 공백 제거

her0807 committed:
fix : 당첨결과 정렬

her0807 committed:
refactor : 출력문 예시와 동일하게 수정

her0807 committed:
refactor : 코드 정리(컨벤션,메서드,테스트)

her0807 committed:
refactor :  패키지 구조 변경, 싱글톤 적용, view 유틸클래스로 변경

- 패키지 구조에 service 생성
- LottoMachine 을 싱글톤으로 변경
- InputView 를 유틸클래스로 변경

her0807 committed:
refactor : controller 메서드 분리

her0807 committed:
refactor : List 반환 시 unmodifiableList() 로 변경

her0807 committed:
docs : 예외사항 추가 , 페어 룰 변경

her0807 committed:
refactor : test 메서드 명 영어로 변경

her0807 committed:
docs : 구입금액에 대한 예외 사항 수정

her0807 committed:
fix : 구입 금액 검증 로직 변경

구입 금액이 1000원 단위로 나뉘냐가 아니라, 1000원 이상인지로 변경

her0807 committed:
refactor : 도메인과 뷰 의존성 제거

her0807 committed:
refactor : 당첨 갯수 계산 로직 변경

her0807 committed:
test : lottos 에서 당첨 결과가 nothing 일경우 무시하는지?

her0807 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,32 @@
+	private InputView() {
+	}
+
+	public static String insertPayment() {
+		OutputView.printGuideMessage("구입금액을 입력해 주세요.");
+		return scanner.nextLine();
```
## View 에서 객체를 반환하는 것에 대하여.
객체 생성의 책임을 view 에서 하면 controller 에서 하는 일을 조금 줄일 수 있겠다는 단순한 생각에
이 부분의 원래 코드는 객체를 반환했었는데요, 

```java

	public static Payment insertPayment() {
		try {
			OutputView.printGuideMessage("구입금액을 입력해 주세요.");
			return new Payment(scanner.nextLine());
		} catch (Exception e) {
			OutputView.printErrorMessage(e.getMessage());
			return insertPayment();
		}
	}

```

다른 크루와 함께 코드를 보다가  `뷰에서 도메인을 알면 안된다. ` 라는 말을 들었습니다. 
그리고 다시 바라보니 현재는 콘솔로 입력을 받고 있지만, 웹 서비스를 하는 관점에서 봤을 때는
request 가 controller 에 날아오는 시점에는 json 형식으로 넘어오고, 그 값을 파싱하는 것은 컨트롤러였던게 생각이나서
현재 코드로 바꾸었습니다.  

그런데 명확하게 와닿지는 않는 상태라 객체를 반환하는 것에 대하여 왜 안 좋은가?
핀의 의견을 듣고 싶습니다 !!!

lalize replied:
어떤 이유로 `뷰에서 도메인을 알면 안된다.`라는 결론을 내리게 되었는지 궁금하네요!

제가 생각하기로는 View에서 의도하지 않은 방법으로 도메인 객체가 사용될 수 있고, View를 수정할 때 도메인도 함께 수정하게 될 수도 있을 것 같아요. 반대로 도메인이 수정되었을 때 View도 함께 수정해야하는 경우도 생길 것 같구요. 변경 포인트를 최소화 하는 게 좋지 않을까요?

여기서 질문 하나 드릴게요. 아래 생성자는 어떤 의도로 생성되었나요?
```java
public Payment(String payment) {
...
}
```

her0807 replied:
@lalize  입력값에 대한 검증을 모두 도메인 내부에서 하고 싶었습니다!
그래서 입력 받는 가장 기본 값인 String 을 받아서 int 에 관한 처리도 내부에서 해주었습니다.

her0807 replied:
추가로 생각해보니.. 상태 타입이 Int 이기 때문에 주 생성자 타입은 int 로 처리하는게 더 자연스럽고,
string 으로 들어오는 경우는 서브 생성자가 처리해주는게 좋겠네요 :)

her0807 commented on src/main/java/service/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+
+	public static LottoMachine getInstance() {
+		return INSTANCE;
+	}
```
## 싱글톤과 유틸 클래스의 차이

이 부분은 페어와 의견을 나누다가 싱글톤을 사용하게 되었습니다.

> #### 싱글톤은 
하나의 객체를 공유한다. 객체 내부의 상태가 있을 때 상태가 변해도 공유하는 모든 것들에게 그 상태를 보장한다. 라고 생각합니다. 

**장점**
- 상태를 보장한다.
- 중복 사용이 가능하다. 

> #### 유틸클래스
불변하는 상수들을 모아두거나, 행위만 있을 경우 사용을 한다고 생각합니다.

**장점**
- 객체 생성을 하지 않아도 되어, `클래스명.메서드명() ` 과 같이 바로 사용이 가능해 사용성이 증가한다. 
- 중복 사용이 가능하다. 

---

### 질문
현재 상태에서는 1 ~45 의 범위에서 로또 번호를 뽑지만, 그 번호가 바뀔 수도 있고,
수동 로또가 추가될 수도 있어서 그런 것들을 고려했을 때 
싱글톤으로 해보면 좋겠다고 생각되어 이렇게 코드를 작성했습니다.

**핀이 생각하기에 싱글톤은 어떨 때 사용하면 좋을지, 현재 잘 생각하고 사용한게 맞는지 궁금합니다.** 🤕

lalize replied:
수달이 학습하신 내용에 나와 있는 것처럼 클래스의 객체가 프로그램 전체 영역에서 단 하나만 존재해야 하고 상태와 행위를 공유해야 하는 경우가 생길 때 싱글톤을 사용하면 되지 않을까요?

로또 번호의 정책이 바뀌고 수동 로또 기능이 생길 때 싱글톤을 어떻게 사용하게 될지 조금 더 자세히 알려주실 수 있으실까요?
단순히 질문하신 내용만 봤을 때는 로또 정책이 바뀌었을 때 싱글톤이 어떤 문제를 해결해주길 기대하는 건지 파악이 어려워요 😢

her0807 replied:
@lalize 제가 맨처음에 이 클래스를 작성했을 때는  로또 생성에 관한 내용을 전부 이 객체에 두고 싶었습니다. 그리고 객체를 하나만 생성해서 사용하고 싶어서 유틸과 싱글톤 중 하나를 써보자! 했는데 유틸은 자주 사용했지만, 싱글톤은 사용해본 적이 없어서 한번 적용해보고 싶었습니다. 

그래서 수동과 자동 모두 이 객체가 하게끔.. 하고 싶었는데 생각해보니까 행위는 있지만, 상태가 없어서 싱글톤으로 해야할 이유가 없네요 .. !! ^^ 리팩토링 하겠습니다..

her0807 commented on src/main/java/domain/Payment.java:
```diff
@@ -0,0 +1,45 @@
+	public double calculateProfitRate(int totalProfit) {
+		return (double)totalProfit / payment;
+	}
+}
```
## 수익률 계산이 구입금액의 역할인가?
이 부분은 사실, 수익률을 계산하는건 LottoResult 의 역할 같은데
 getPayment() 를 사용하고 싶지 않아서 이렇게 작성하게 되었습니다. 더 좋은 방법이 있을까요?

lalize replied:
totalProfit은 LottoResult가 알고 있고 payment는 Payment가 알고 있으니 이 둘을 협력 관계로 만들면 getter 없이 할 수 있지 않을까요?
힌트: `LottoResult::calculateProfitRate(Payment payment)` 메서드 시그니처를 만들어보기!

her0807 replied:
payment 와 lottoResult 의 협력 관계를 고민해보겠습니다 :)

her0807 committed:
refactor : 컨벤션에 맞추어 수정

lalize commented on README.md:
```diff
@@ -1,6 +1,76 @@
+- **수익률을 계산한다**
+    - (당첨금액 / 투입금액)
+
+## 페어 프로그래밍 룰
```
👍

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,63 @@
+
+public class Lotto {
+	private static final int LOTTO_SIZE = 6;
+	private final List<LottoNumber> lotto;
```
상수와 필드 사이 빈줄을 추가하여 가독성을 높여볼까요? 😃

her0807 replied:
앗 .. 가독성 ++ 해보겠습니다. !!!!!

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,63 @@
+		checkLottoSize(lottoNumbers);
+		checkDuplicatedLottoNumber(lottoNumbers);
+		this.lotto = lottoNumbers;
+	}
```
주 생성자를 통해 직접 객체를 생성하면 외부에서 lottoNumbers를 변경할 수 있겠네요~

her0807 replied:
변경 가능성에 대해서는 생각하지 못했네요 🤕  ㅠㅠ [방어적 복사](https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/) 키워드를 참고하여 리팩토링 해보겠습니다!

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,63 @@
+
+	private static LottoNumber createLottoNumber(String number) {
+		return new LottoNumber(number);
+	}
```
필요한가요?

her0807 replied:
흡 ... !

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,63 @@
+	@Override
+	public String toString() {
+		return "추첨된 번호는 " + lotto + " 입니다.";
+	}
```
toString을 재정의한 이유가 View를 위한거면 직접적인 의존관계는 없어도 도메인이 View를 의존하고 있다고 볼 수 있지 않을까요?

her0807 replied:
@lalize  제가 내용 찍어볼 때 친절하게 보고 싶어서 .. 저렇게 작성했습니다.. 🥰

lalize commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,54 @@
+public class LottoNumber implements Comparable<LottoNumber> {
+	private static final int MIN_BOUND = 1;
+	private static final int MAX_BOUND = 45;
+	private int lottoNumber;
```
final을 붙여 내부에서도 상태가 변경되지 않도록 할까요?

lalize commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,54 @@
+	public LottoNumber(int lottoNumber) {
+		checkRange(lottoNumber);
+		this.lottoNumber = lottoNumber;
+	}
```
생성자마다 중복으로 검증을 하고 있는데요!
하나의 생성자를 주 생성자로 만들고 나머지 생성자는 부 생성자로 하면 어떨까요?
```java
// 부 생성자
public Abc(String abc) {
  this(Integer.parseInt(abc));
}

// 주 생성자
public Abc(int abc) {
  check(abc);
  this.abc = abc;
}
```

her0807 replied:
와! 생성자에서 생성자를 호출할 수 있는지 처음 알았습니다! ㅎㅎㅎ 공부해보겠습니다.

lalize commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,54 @@
+	@Override
+	public int compareTo(LottoNumber o) {
+		return this.lottoNumber - o.lottoNumber;
+	}
```
`Integer.compare`라는 안정된 API를 이용하여 compareTo를 구현하는 건 어떨까요?

```suggestion
	@Override
	public int compareTo(LottoNumber o) {
		return Integer.compare(this.lottoNumber, o.lottoNumber);
	}
```

lalize commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,27 @@
+
+	public LottoResult(Map<Rank, Integer> ranks) {
+		this.ranks = ranks;
+	}
```
ranks가 null로 입력되면 어떻게 될까요?

her0807 replied:
나중에 사용할 때 에러가 나서 에러를 찾기 어렵겠네요! [매개변수 Null 검증](https://umbum.dev/1029)을 참고하여 수정했습니다.

lalize commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,27 @@
+
+	private int calculateRankProfit(Rank rank, int count) {
+		return rank.getMoney() * count;
+	}
```
값을 꺼내지 않고 rank에게 메시지를 보낼 수도 있겠네요!

her0807 replied:
enum 클래스가 가지고 있는 값들은 상수라고 생각해서 get 했었는데, 메세지를 보내는게 좋겠네요 핀 감사합니다. :)

lalize commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+		lottos.stream()
+			.map(lotto -> winningLotto.calculateRank(lotto))
+			.filter(rank -> !rank.isNothing())
+			.forEach(rank -> rankCounts.replace(rank, rankCounts.get(rank) + 1));
```
`Collectors.groupingBy`를 학습해보시면 좋을 것 같아요!

lalize commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+			.map(lotto -> winningLotto.calculateRank(lotto))
+			.filter(rank -> !rank.isNothing())
+			.forEach(rank -> rankCounts.replace(rank, rankCounts.get(rank) + 1));
+		return Collections.unmodifiableMap(rankCounts);
```
객체 내부 상태를 반환하는 게 아닌 새 맵을 만들어 반환하는데 `unmodifiableMap`를 해줄 필요가 있을까요?

her0807 replied:
다음부터는 의식적으로 작성하겠습니다. 감사합니다 핀 🤓

lalize commented on src/main/java/domain/Payment.java:
```diff
@@ -0,0 +1,45 @@
+		checkNegative(changeInt);
+		checkPaymentLimit(changeInt);
+		return changeInt;
+	}
```
int를 받는 주 생성자를 두고 해당 코드는 부 생성자로 두면 클래스가 조금 더 유연해질 것 같아요!

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public static Map<Rank, Integer> getMap() {
+		final Map<Rank, Integer> rankMap = new LinkedHashMap<>();
```
[EnumMap](https://www.baeldung.com/java-enum-map)을 학습해보시면 좋을 것 같아요!

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,56 @@
+		final Map<Rank, Integer> rankMap = new LinkedHashMap<>();
+		Arrays.stream(values())
+			.filter(rank -> !rank.equals(NOTHING))
+			.forEach(rank -> rankMap.put(rank, 0));
```
getMap은 LottoResult의 책임에 가까운 것 같은 느낌적인 느낌이 드네요 🤔

her0807 replied:
단순히 Rank 를 Map 으로 반환 한다고 생각하여 Rank 로 위치해 두었는데,
핀 말씀을 듣고 나니 LottoResult를 위한 작업이고, filter 로 전체 Rank 가 아닌 nothing 을 제거하는 부분도 있어
LottoResult 에 static 으로 존재하는게 책임이 더 명확해지는 것 같네요!

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,56 @@
+
+	public boolean isNothing() {
+		return this.equals(NOTHING);
+	}
```
Enum 비교는 동일성 비교로 하면 더 좋습니다!
```suggestion
	public boolean isNothing() {
		return this == NOTHING;
	}
```

her0807 replied:
동일성과 동등성 키워드로 자세히 공부해보겠습니다!

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,56 @@
+	public static Rank getRank(int target, boolean bonus) {
+		return Arrays.stream(Rank.values())
+			.filter(rank -> rank.match == target)
+			.filter(rank -> rank.bonus == bonus)
```
로또 번호 3개, 4개, 6개 맞고 보너스볼까지 맞으면 어떻게 될까요?

her0807 replied:
헉 제대로 된 결과 값이 출력되지 않겠군요 .. 조건문에 대해서 조금더 깊게 생각해보겠습니다!

lalize commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,22 @@
+package domain;
+
+public class WinningLotto {
+	private final Lotto lotto;
+	private final LottoNumber bonusNumber;
```
💯

her0807 replied:
상속보다 조합을 !!! 💇🏻‍♀️

lalize commented on src/main/java/service/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+	static {
+		IntStream.rangeClosed(MIN_BOUND, MAX_BOUND)
+			.mapToObj(LottoNumber::new)
+			.forEach(LOTTO_BUCKET::add);
```
로또 번호는 1~45의 숫자만 가질 수 있으니 로또 번호 클래스가 자체적으로 로또 번호를 캐싱하고 재사용하면 어떨까요?

* [정적 팩토리 메서드(Static Factory Method)는 왜 사용할까?](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)

her0807 replied:
정적 팩토리 메서드를 사용해보자 ~ 정도만 생각하고 구현했던 부분인데, 재사용도 할 수 있다니 ..
아주 좋은 방향입니다. 변경해보겠습니다. 핀 👍🏻 👍🏻

lalize commented on src/main/java/service/LottoMachine.java:
```diff
@@ -0,0 +1,47 @@
+	private Lotto createAutoLotto() {
+		Collections.shuffle(LOTTO_BUCKET);
+		return new Lotto(LOTTO_BUCKET.stream()
+			.limit(6)
```
매직넘버는 상수화 해주세요 😃

her0807 replied:
헙 😵‍💫  시정하겠습니다!!!

lalize commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,45 @@
+	void range_min() {
+		assertThatThrownBy(() -> new LottoNumber("0"))
+			.isInstanceOf(IllegalArgumentException.class);
+	}
```
경계값 테스트 💯 
`@ParameterizedTest` 에 대해 학습도 해보시면 좋을 것 같아요!

lalize commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,70 @@
+			.doesNotThrowAnyException();
+	}
+
+	@DisplayName("개수 실패")
+	@Test
+	void lotto_size_fail() {
+		assertThatThrownBy(() -> new Lotto(lottos))
+			.isInstanceOf(IllegalArgumentException.class);
+	}
+
+	@DisplayName("중복 금지")
+	@Test
+	void duplicate_fail() {
+		lottos.add(new LottoNumber(5));
+		assertThatThrownBy(() -> new Lotto(lottos))
+			.isInstanceOf(IllegalArgumentException.class);
+	}
```
어떤 검증이 실패해서 예외가 발생했는지 알 수 있나요?

her0807 replied:
hasMessage() 를 통해서 구체적으로 어떤 예외인지 확인해보겠습니다!

lalize commented on src/test/java/domain/PaymentTest.java:
```diff
@@ -0,0 +1,81 @@
+	void range_max() {
+		assertThatThrownBy(() -> new Payment("110000"))
+			.isInstanceOf(IllegalArgumentException.class);
+	}
```
10만원을 초과해서 실패한건지, 1000원 미만이여서 실패한건지 등 테스트로 확인이 되나요?

her0807 replied:
hasMessage() 를 통해서 구체적으로 어떤 예외인지 확인해보겠습니다!

lalize reviewed (changes requested):
안녕하세요 수달! 🦦 
작성하신 코드 잘 확인했습니다 👍 
몇 가지 코멘트와 질문에 대한 답변을 남겨두었으니 확인 부탁드릴게요!
추가로 궁금한 점 있으시면 코멘트나 DM 편히 주세요!

her0807 committed:
refactor :  구체적인 예외 검증

her0807 committed:
refactor : payment 주 생성자 타입 변경

her0807 committed:
refactor : payment 와 LottoResult 협력 관계 생성

her0807 committed:
refactor : 매직 넘버 수정

her0807 committed:
refactor : 정적 팩토리 메서드 LottoNumber 에 적용

her0807 committed:
refactor : LottoResult null 체크 / LottoNumber compareTo 재정의

her0807 committed:
refactor : rank get -> 메세지 보내기로 수정

her0807 committed:
refactor : lottos 의 countRank 부분 개선

- forEach 문으로 Map 의 값을 갱신 해주면서 count 를 했던 부분을
groupingBy 를 사용하여 계산 후, 전체 rank 를 가지고 있는 ranks 와 merge 하는 형식으로 변경

her0807 committed:
refactor : Map -> EnumMap 으로 변경

her0807 committed:
fix : Rank of 조건 오류 수정

her0807 committed:
refactor :  컨벤션에 맞추어 수정

her0807 committed:
refactor : 불필요한 메서드 삭제

her0807 committed:
refactor : null에 대한 검증 처리

her0807 committed:
refactor : 객체 생성 로직 분리

her0807 committed:
refactor : 생성자 재사용

lalize commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,24 @@
+
+public class LottoController {
+
+	public void run() {
+		Payment payment = InputConvertor.createPayment();
+		int lottoCount = payment.calculateLottoCount();
+		OutputView.printLottoCount(lottoCount);
+
+		Lottos lottos = InputConvertor.createLottos(lottoCount);
+		OutputView.printLottos(lottos);
+
+		WinningLotto winningLotto = InputConvertor.createWinningLotto();
+
+		OutputView.printLottoResult(new LottoResult(lottos.countRank(winningLotto)), payment);
+	}
```
코드가 굉장히 읽기 쉬워졌네요!! 💯

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+public class Lotto {
+	public static final int LOTTO_SIZE = 6;
+
+	private final List<LottoNumber> lotto;
```
`List`보다 더 적합한 타입이 있을 것 같아요 ㅎㅎ

her0807 replied:
핀의 피드백을 보고 Lotto가 가지고 있는 조건을 다시 생각해보았습니다.
로또는 숫자 6개를 가지며, 중복을 허용하지 않는다. 라는 조건이 있었고, 그 내용을 바탕으로 List, Set 이 떠올랐습니다. 

- **List** 는 중복을 허용하고 순서를 보장하며 데이터를 보관할 때 사용된다.

- **Set** 는 중복을 허용하지 않고 순서도 보장할 필요 없이 데이터를 보관할 때 사용된다.

로또 번호는  중복을 허용하지 않으니 Set 이 더 적절하다는 결론을 내렸고,
그 중에서도 오름차순으로 정렬해주는 treeSet 이 떠올랐습니다 👍🏻

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+			int lottoNumber = Integer.parseInt(number.trim());
+			lotto.add(LottoNumber.of(lottoNumber));
+		}
+		return Collections.unmodifiableList(lotto);
```
여기서는 `unmodifiableList`를 하지 않아도 되지 않을까요?

her0807 replied:
여기서 unmodifiableList 로 반환한 뒤 또 한번 생성자에서 방어적 복사를 또 하니 핀 말씀처럼  불필요할 것 같네요. ㅎㅎ

lalize replied:
toLotto 메서드 첫 줄에서 new ArrayList<>(); 를 하여 새 리스트를 생성하기 때문에 드렸던 코멘트 입니다:)

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+			.count();
+	}
+
+	public boolean isContain(LottoNumber lottoNumber) {
```
`contains`라고 해도 될 것 같아요 ㅎㅎ

lalize commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	public List<LottoNumber> getLotto() {
+		return Collections.unmodifiableList(lotto);
```
getter는 방어적 복사가 아닌 unmodifiableList를 사용한 이유가 궁금하네요!

her0807 replied:
@lalize 
제가 이해했던 두개의 개념은 아래와 같습니다. 
- **방어적 복사**는 복사한 값이 변화해도, 원본에 영향을 미치지 않는다. 
- **unmodifiableList**는 값을 읽는 것 외에 다른 연산을 시도할 경우 예외처리가 발생한다. 

getter 는 view 에서 값을 읽는 것 외에 다른 용도로는 사용하고 싶지 않아 연산 시도시 에러가 발생했으면 좋겠다고 생각하여
unmodifiableList 로 작성했습니다!

lalize commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,75 @@
+	private static final int MIN_BOUND = 1;
+	private static final int MAX_BOUND = 45;
+
+	private static Map<Integer, LottoNumber> LOTTO_NUMBERS = new HashMap<>();
+
+	static {
+		IntStream.rangeClosed(MIN_BOUND, MAX_BOUND)
+			.forEach(number -> LOTTO_NUMBERS.put(number, new LottoNumber(number)));
+	}
```
로또 번호 캐싱 💯

lalize commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,75 @@
+
+	private final int lottoNumber;
+
+	public LottoNumber(int lottoNumber) {
+		checkRange(lottoNumber);
+		this.lottoNumber = lottoNumber;
+	}
+
+	public LottoNumber(String lottoNumber) {
+		this(Integer.parseInt(lottoNumber));
+	}
```
로또 번호는 항상 캐싱된 것만 사용할 수 있도록 public 생성자를 private으로 변경하는 건 어떨까요?

her0807 replied:
오! 핀! 이 부분은 피드백 보기 전에 떠올랐던 부분인데 말씀해 주시니 반갑네요 🥰

lalize commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	public static EnumMap<Rank, Long> getRankMap() {
+		EnumMap<Rank, Long> ranks = new EnumMap<Rank, Long>(Rank.class);
```
```suggestion
		Map<Rank, Long> ranks = new EnumMap<Rank, Long>(Rank.class);
```

lalize commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,33 @@
+	private final List<Lotto> lottos;
+
+	public Lottos(List<Lotto> lottos) {
+		this.lottos = Objects.requireNonNull(lottos,"[ERROR] lottos가 null 입니다.");
```
어떤 부분은 방어적 복사를 하고 어떤 부분은 입력 받은 값을 그대로 초기화하고 있는데요~
수달은 방어적 복사를 어떤 기준으로 하는 지 얘기해 주실 수 있으실까요?

her0807 replied:
제가 아직 방어적 복사를 하지 않아 발생된 에러(?) 에 맞아본 적이 없어서 그런지 자꾸만 놓치네요 😮‍💨 

Objects.requireNonNull() 후 바로 값을 반환하여 초기화 하면 깔끔하니 좋겠다라고 
생각만 하고, 방어적 복사는 고려하지 않았던 것 같습니다. (반복되는 피드백 죄송합니다 😿 )

lalize commented on src/main/java/domain/Payment.java:
```diff
@@ -0,0 +1,48 @@
+package domain;
+
+public class Payment {
+	private static final int LOTTO_PRICE = 1000;
+	private static final int PAYMENT_LIMIT_PRICE = 100000;
+	private final int payment;
```
이전에 드렸던 피드백입니다!
https://github.com/woowacourse/java-lotto/pull/381#discussion_r814349102

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,58 @@
+
+public enum Rank {
+	NOTHING(0, 0, false),
+	FIFTH(5000, 3, false),
+	FOURTH(50000, 4, false),
+	THIRD(150000, 5, false),
+	SECOND(30000000, 5, true),
+	FIRST(2000000000, 6, false);
```
`_`를 추가하면 가독성에 더 좋아요!
```suggestion
	FIFTH(5_000, 3, false),
	FOURTH(50_000, 4, false),
	THIRD(150_000, 5, false),
	SECOND(30_000_000, 5, true),
	FIRST(2_000_000_000, 6, false);
```

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,58 @@
+			.orElse(NOTHING);
+	}
+
+	private boolean isSameRank(int matchCount) {
+		return this.match == matchCount;
+	}
```
2, 3등 때문에 메서드명이 적절하지 않은 것 같아요!
그리고 해당 메서드가 내부 구현에만 사용되고 있는데요~ 메서드명과 구현이 동일한 추상화 수준인 것 같은데 메서드로 추출해서 어떤 이점이 있었나요?

her0807 replied:
추후 재사용할 수도 있겠다는 생각이 들어 메서드로 추출했었는데, 현재 상태에서는 메서드로 추출하지 않아도 괜찮겠네요 .

lalize commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,58 @@
+		return match;
+	}
+
+	public Rank getRank() {
+		return this;
+	}
```
사용하는 쪽에서는 Rank를 아는데 이게 필요한가요?!

her0807 replied:
groupingBy(Rank.getRank) 를 사용하려고 작성했던 코드인데, Function.identity()로 변경하면 getter 없이 값을 가져올 수 있음을 알게되어 변경하도록 하겠습니다 👍🏻

lalize commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,28 @@
+	private final LottoNumber bonusNumber;
+
+	public WinningLotto(Lotto lotto, LottoNumber bonusNumber) {
+		this.lotto = Objects.requireNonNull(lotto, "[ERROR] WinningLotto에 lotto가 null 입니다.");
+		this.bonusNumber = Objects.requireNonNull(bonusNumber, "[ERROR] WinningLotto에 bonusNumber기 null 입니다.");
+		checkDuplicated(lotto, bonusNumber);
```
할당 전에 검증하면 어떨까요?

lalize commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,48 @@
+	@DisplayName("숫자를 입력했을 시 성공")
+	@Test
+	void input_data_success() {
+		assertThatCode(() -> new LottoNumber("45"))
```
1에 대해서도 테스트해볼까요?

lalize commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,70 @@
+		assertThat(count).isEqualTo(5);
+	}
+
+	@AfterEach
+	void clear() {
+		lottos.clear();
+	}
```
BeforeEach에서 add하지 않고 리스트를 초기화 해주는 게 어떨까요?

lalize reviewed (approved):
수달 🦦  리뷰 반영 잘해주셨네요~ 👍 
1단계는 여기서 머지해도 될 것 같아요~
몇 가지 코멘트 남겨두었으니 확인 부탁드릴게요.
다음 단계 진행하면서 리뷰 반영도 같이 해주시면 좋을 것 같아요!
추가로 궁금한 점 있으시면 코멘트나 DM 편하게 주세요 ~~
